File: mira-hub/src/lib/proposal-transition.ts
/**
 * proposal-transition.ts — the single Hub-side writer for proposal status.
 *
 * Implements ADR-0017 ("Proposal state-machine mapping — one logical machine,
 * three table projections"). One logical decision fans out to three tables,
 * each with its own enum:
 *
 *   ai_suggestions.status          pending|accepted|rejected|deferred|superseded
 *   relationship_proposals.status  proposed|reviewed|verified|rejected|deprecated|contradicted
 *   kg_*.approval_state            proposed|verified|rejected|needs_review   (engine side, Python)
 *
 * This module owns the FIRST TWO (Hub) projections. kg_*.approval_state is the
 * engine's hot-path read and is written by mira_bots/shared/proposal_transition.py
 * and by the decide route's existing kg_relationships upsert.
 *
 * Per .claude/CLAUDE.md § "Knowledge graph proposals": once this helper exists,
 * a direct `UPDATE … SET status = …` on ai_suggestions / relationship_proposals
 * is a bug — go through applyHubProposalTransition().
 *
 * The mapping table below is the load-bearing artifact (ADR-0017 §Decision).
 */

export type ProposalTrigger =
  | "new" // a fresh LLM proposal lands
  | "accept" // admin accepts on Hub
  | "reject" // admin rejects on Hub
  | "defer" // admin punts ("ask me later")
  | "supersede" // a newer proposal replaces this one
  | "contradict" // engine finds contradicting evidence
  | "flag_review"; // engine flags an edge for human re-look

export interface HubStatusMap {
  /** ai_suggestions.status, or null to leave unchanged. */
  aiSuggestion: string | null;
  /** relationship_proposals.status (kg_edge only), or null to leave unchanged. */
  relationshipProposal: string | null;
}

/** Canonical ADR-0017 mapping (Hub-side columns only). */
export const PROPOSAL_TRANSITIONS: Record<ProposalTrigger, HubStatusMap> = {
  new: { aiSuggestion: "pending", relationshipProposal: "proposed" },
  accept: { aiSuggestion: "accepted", relationshipProposal: "verified" },
  reject: { aiSuggestion: "rejected", relationshipProposal: "rejected" },
  defer: { aiSuggestion: "deferred", relationshipProposal: null },
  supersede: { aiSuggestion: "superseded", relationshipProposal: "deprecated" },
  contradict: { aiSuggestion: "pending", relationshipProposal: "contradicted" },
  flag_review: { aiSuggestion: "pending", relationshipProposal: "reviewed" },
};

/** Minimal query interface — matches the pg client passed by withTenantContext. */
export interface QueryClient {
  query: (text: string, params?: unknown[]) => Promise<{ rows: unknown[] }>;
}

export interface ProposalTransitionInput {
  trigger: ProposalTrigger;
  /** relationship_proposals.id (for kg_edge proposals). */
  relationshipProposalId?: string;
  /** ai_suggestions.id, if known directly. */
  aiSuggestionId?: string;
  /** Audit label, e.g. `human:<userId>`. */
  reviewerLabel?: string;
  /** Optional reason, stored on relationship_proposals.reasoning. */
  reason?: string;
}

export interface ProposalTransitionResult {
  trigger: ProposalTrigger;
  relationshipProposalStatus: string | null;
  aiSuggestionStatus: string | null;
  aiSuggestionRowsUpdated: number;
}

/**
 * Apply the ADR-0017 Hub-side projections for one logical decision, inside the
 * caller's transaction. Updates relationship_proposals (when an id + mapped
 * status are present) AND the paired ai_suggestions row (matched by
 * payload->>'relationship_proposal_id', or by explicit aiSuggestionId) so the
 * two Hub projections never drift apart — the gap the ADR-0017 canary detects.
 *
 * Engine-side kg_*.approval_state is NOT written here (Python helper / the
 * decide route's kg_relationships upsert own it).
 */
export async function applyHubProposalTransition(
  client: QueryClient,
  input: ProposalTransitionInput,
): Promise<ProposalTransitionResult> {
  const map = PROPOSAL_TRANSITIONS[input.trigger];
  if (!map) {
    throw new Error(`unknown proposal trigger: ${String(input.trigger)}`);
  }
  const reviewer = input.reviewerLabel ?? null;
  const reason = input.reason ?? "";

  // 1) relationship_proposals (kg_edge projection)
  if (map.relationshipProposal && input.relationshipProposalId) {
    await client.query(
      `UPDATE relationship_proposals
          SET status = $1,
              reviewed_at = now(),
              reviewed_by = COALESCE($2, reviewed_by),
              reasoning = COALESCE(NULLIF($3, ''), reasoning)
        WHERE id = $4`,
      [map.relationshipProposal, reviewer, reason, input.relationshipProposalId],
    );
  }

  // 2) ai_suggestions (Hub queue projection) — keep the paired row in lockstep.
  // The kg_edge AISuggestion points at its relationship_proposals row via the
  // JSONB column `extracted_data` -> 'relationship_proposal_id' (verified
  // against mig 027 + the ADR-0017 canary, which keys off the same column —
  // NOT `payload`).
  let aiRows = 0;
  if (map.aiSuggestion) {
    if (input.aiSuggestionId) {
      const r = await client.query(
        `UPDATE ai_suggestions
            SET status = $1, updated_at = now(),
                reviewed_by = COALESCE($2, reviewed_by),
                reviewed_at = now()
          WHERE id = $3`,
        [map.aiSuggestion, reviewer, input.aiSuggestionId],
      );
      aiRows = (r as { rowCount?: number }).rowCount ?? r.rows.length;
    } else if (input.relationshipProposalId) {
      const r = await client.query(
        `UPDATE ai_suggestions
            SET status = $1, updated_at = now(),
                reviewed_by = COALESCE($2, reviewed_by),
                reviewed_at = now()
          WHERE suggestion_type = 'kg_edge'
            AND NULLIF(extracted_data->>'relationship_proposal_id', '')::uuid = $3::uuid`,
        [map.aiSuggestion, reviewer, input.relationshipProposalId],
      );
      aiRows = (r as { rowCount?: number }).rowCount ?? r.rows.length;
    }
  }

  return {
    trigger: input.trigger,
    relationshipProposalStatus: map.relationshipProposal,
    aiSuggestionStatus: map.aiSuggestion,
    aiSuggestionRowsUpdated: aiRows,
  };
}

File: mira-bots/shared/proposal_transition.py
"""proposal_transition.py — engine-side writer for KG approval state (ADR-0017).

ADR-0017 ("Proposal state-machine mapping — one logical machine, three table
projections") keeps three enums distinct and centralizes the writes. This module
owns the ENGINE-SIDE projection — ``kg_entities.approval_state`` /
``kg_relationships.approval_state`` (and the engine-triggered transitions on
``relationship_proposals``). The Hub-side projections (``ai_suggestions`` /
``relationship_proposals`` from admin actions) are owned by the TypeScript
counterpart ``mira-hub/lib/proposal-transition.ts``.

Per .claude/CLAUDE.md § "Knowledge graph proposals": once this helper exists, a
direct ``UPDATE … SET approval_state = …`` that bypasses it is a bug.

The mapping table below is the load-bearing artifact (ADR-0017 §Decision). It is
intentionally pure/data-only so it can be imported and unit-tested without a DB.
"""

from __future__ import annotations

from typing import Optional

# Triggers shared with the TS helper (mira-hub/lib/proposal-transition.ts).
TRIGGERS = ("new", "accept", "reject", "defer", "supersede", "contradict", "flag_review")

# Canonical ADR-0017 mapping, ENGINE-SIDE columns only.
#   kg_approval_state          → kg_entities / kg_relationships approval_state
#   relationship_proposal      → engine-triggered relationship_proposals.status
# None means "leave unchanged".
PROPOSAL_TRANSITIONS: dict[str, dict[str, Optional[str]]] = {
    "new": {"kg_approval_state": "proposed", "relationship_proposal": "proposed"},
    "accept": {"kg_approval_state": "verified", "relationship_proposal": "verified"},
    "reject": {"kg_approval_state": "rejected", "relationship_proposal": "rejected"},
    "defer": {"kg_approval_state": None, "relationship_proposal": None},
    "supersede": {"kg_approval_state": None, "relationship_proposal": "deprecated"},
    # Engine job finds contradicting evidence: a verified edge goes back to review.
    "contradict": {"kg_approval_state": "needs_review", "relationship_proposal": "contradicted"},
    # Engine flags an edge for a human re-look.
    "flag_review": {"kg_approval_state": "needs_review", "relationship_proposal": "reviewed"},
}

# Valid target values per column (defense: never emit a value a reader rejects).
KG_APPROVAL_STATES = frozenset({"proposed", "verified", "rejected", "needs_review"})
RELATIONSHIP_PROPOSAL_STATES = frozenset(
    {"proposed", "reviewed", "verified", "rejected", "deprecated", "contradicted"}
)


def kg_approval_for(trigger: str) -> Optional[str]:
    """The kg_*.approval_state value for a trigger (None = unchanged)."""
    if trigger not in PROPOSAL_TRANSITIONS:
        raise ValueError(f"unknown proposal trigger: {trigger!r}")
    return PROPOSAL_TRANSITIONS[trigger]["kg_approval_state"]


def relationship_proposal_status_for(trigger: str) -> Optional[str]:
    """The engine-triggered relationship_proposals.status (None = unchanged)."""
    if trigger not in PROPOSAL_TRANSITIONS:
        raise ValueError(f"unknown proposal trigger: {trigger!r}")
    return PROPOSAL_TRANSITIONS[trigger]["relationship_proposal"]


def apply_kg_approval(cur, *, table: str, row_id: str, trigger: str, tenant_id: str) -> bool:
    """Set ``approval_state`` on a kg_entities / kg_relationships row per the
    ADR-0017 mapping, using the caller's DB cursor (psycopg/SQLAlchemy-style
    ``cur.execute(sql, params)``). Returns True if a write was issued.

    ``table`` must be 'kg_entities' or 'kg_relationships' (validated — the value
    is interpolated into the statement, so it is NOT free-form user input)."""
    if table not in ("kg_entities", "kg_relationships"):
        raise ValueError(f"table must be kg_entities|kg_relationships, got {table!r}")
    target = kg_approval_for(trigger)
    if target is None:
        return False
    if target not in KG_APPROVAL_STATES:  # pragma: no cover - guarded by mapping
        raise ValueError(f"refusing to write unknown approval_state {target!r}")
    cur.execute(
        f"UPDATE {table} SET approval_state = %s WHERE id = %s AND tenant_id = %s",
        (target, row_id, tenant_id),
    )
    return True
